### PR TITLE
 ActiveRecord::Base#wait deadlock fix, part 2

### PR DESCRIPTION
**Pull request by lowjoel**
I've identified two more contributing factors, in addition to those in #721:
 - Exceptions in threads are swallowed by default, see Thread#abort_on_exception; this can cause jobs to die without anyone knowing, especially if the job was not instantiated in the thread yet
 - We spawn jobs within a transaction; this is a problem because the transaction creates the job (in the Jobs table), but the job can start and look for itself in the jobs table *before the transaction is committed*. This causes some jobs to fail without warning, see point 1.

I've also reduced the number of spurious jobs in the jobs table by only committing the job to the table when the job is enqueued, that should reduce the number of blank jobs in the jobs table. This helps debugging.

### CODE REVIEW

File: lib/autoload/active_job/queue_adapters/background_thread_adapter.rb [hunk 1]
```diff
@@ -10,11 +10,13 @@
...
   class << self
     def enqueue(job) #:nodoc:
-      Thread.new do
+      thread = Thread.new do
         ActiveRecord::Base.connection_pool.with_connection do
           ActiveJob::Base.execute(job.serialize)
         end
       end
+
+      thread.abort_on_exception = true
```
Comment (allenwq) [inline lib/autoload/active_job/queue_adapters/background_thread_adapter.rb]: @lowjoel This is for debugging or ?

File: app/models/course/discussion/topic.rb [hunk 2]
```diff
@@ -19,7 +19,9 @@ def subscribed_by?(user)
   # @param [User] user The user who needs to subscribe to this topic
   def ensure_subscribed_by(user)
     return true if self.subscribed_by?(user)
-    subscriptions.build(user: user).save!
+    ActiveRecord::Base.transaction do
+      subscriptions.build(user: user).save!
```
Comment (allenwq) [inline app/models/course/discussion/topic.rb]: why there is a need to wrap this inside a transaction ?

File: lib/autoload/active_job/queue_adapters/background_thread_adapter.rb (hunk 1 above)
Comment (lowjoel) [inline lib/autoload/active_job/queue_adapters/background_thread_adapter.rb]: this is necessary. the default is to silently swallow the exception and stop the thread.

File: app/models/course/discussion/topic.rb (hunk 2 above)
Comment (lowjoel) [inline app/models/course/discussion/topic.rb]: I put a comment up there, right?

Comment (allenwq) [inline app/models/course/discussion/topic.rb]: Where ?  This is not related to job ?

Comment (lowjoel) [inline app/models/course/discussion/topic.rb]: see answer.rb, crap I didn't copy the comment here

Comment (allenwq) [inline app/models/course/discussion/topic.rb]: OK, So even we have the `rescue ActiveRecord::RecordInvalid`, the enclosing transaction will still rollback ?

Comment (lowjoel) [inline app/models/course/discussion/topic.rb]: no, enclosing transaction will continue.

Comment (allenwq) [inline app/models/course/discussion/topic.rb]: Then what's the purpose of adding this transaction ?  I still not very clear.

Comment (lowjoel) [inline app/models/course/discussion/topic.rb]: any errors caused by a statement in a transaction will require the entire transaction to be rolled back.

here we expect the error -- the duplicate key cannot be guarded against. so the give it a transaction to error, we rollback, and the rest of the program continues.
